Significantly improve speed performance and memory consumption for DataFrame Merge procedure

mzasov opened the pull request:
We experienced a very low performance in DataFrame Merge procedure for big data frames with ~20K and ~70K rows with total 20M rows in the result.

The original idea that was implemented in the low performance code:

foreach (var kvp in newOccurrences)
{
    var newValue = kvp.Value.Where(i => occurrences[kvp.Key].Contains(i)).ToArray();
    if (newValue.Any())
    {
        shrinkedOccurences.Add(kvp.Key, newValue);
    }
}

The first optimization is to use hash sets and calculate intersection natively on each iteration:

foreach (var newOccurrence in newOccurrences)
{
    var newOccurrenceKey = newOccurrence.Key;

    var occurrencesHash = new HashSet<long>(occurrences[newOccurrenceKey]);
    var newOccurrencesHash = new HashSet<long>(newOccurrence.Value);

    newOccurrencesHash.IntersectWith(occurrencesHash);

    if (newOccurrencesHash.Any())
    {
        shrinkedOccurences.Add(newOccurrenceKey, newOccurrencesHash);
    }
}

But there is a high memory consumption that we can improve.

We can avoid usage of hash sets when we know that occurrences are the collections of sorted indexes. Knowledge of that gives us the following implementation of two lists intersections:

var i = 0;
var j = 0;

while (i < sortedList1.Count && j < sortedList2.Count)
{
  var item1 = sortedList1[i];

  while (j < sortedList2.Count)
  {
    var item2 = sortedList2[j];

    if (item1 < item2)
    {
      i++;
      break;
    }
    else if (item1 == item2)
    {
      crossing.Add(item1);
      i++;
      j++;
      break;
    }
    else
    {
      j++;
    }
  }
}

In the version with hash sets we consumed ~20Gb of memory in our example, but in the optimized we see that consumed up to ~500Mb of memory. And speed performance also is the best.

Also there is splitted merge procedure to more smaller functions and improved readability of code.

tarekgh commented on src/Microsoft.Data.Analysis/DataFrame.Join.cs:
```diff
@@ -368,22 +403,62 @@ public DataFrame Merge(DataFrame other, string[] leftJoinColumns, string[] right
+            PrimitiveDataFrameColumn<long> mapIndicesRight = isLeftDataFrameRetained ? supplementaryRowIndices : retainedRowIndices;
+
+            // Insert columns from left dataframe (this)
+            for (var i = 0; i < this.Columns.Count; i++)
```
>var [](http://example.com/codeflow?start=17&length=3)

nit: why did you use var and removed the type here? using the type is better here.

mzasov replied:
fixed

tarekgh commented on src/Microsoft.Data.Analysis/DataFrame.Join.cs:
```diff
@@ -368,22 +403,62 @@ public DataFrame Merge(DataFrame other, string[] leftJoinColumns, string[] right
-            //insert columns from right dataframe (other)
-            for (int i = 0; i < other.Columns.Count; i++)
+            // Insert columns from right dataframe (other)
+            for (var i = 0; i < other.Columns.Count; i++)
```
>var [](http://example.com/codeflow?start=17&length=3)

same. please using the type here.

mzasov replied:
fixed

tarekgh commented on src/Microsoft.Data.Analysis/DataFrame.Join.cs:
```diff
@@ -178,89 +182,121 @@ private static HashSet<long> Merge(DataFrame retainedDataFrame, DataFrame supple
             var retainJoinColumns = retainedJoinColumnNames.Select(name => retainedDataFrame.Columns[name]).ToArray();
-            for (long i = 0; i < retainedDataFrame.Columns.RowCount; i++)
+
+            for (var i = 0; i < retainedDataFrame.Columns.RowCount; i++)
```
>var [](http://example.com/codeflow?start=17&length=3)

please return the type back here.

mzasov replied:
fixed

tarekgh commented on src/Microsoft.Data.Analysis/DataFrame.Join.cs:
```diff
@@ -178,89 +182,121 @@ private static HashSet<long> Merge(DataFrame retainedDataFrame, DataFrame supple
-
-
-            for (int colNameIndex = 0; colNameIndex < retainedJoinColumnNames.Length; colNameIndex++)
+            for (var colNameIndex = 0; colNameIndex < retainedJoinColumnNames.Length; colNameIndex++)
```
>var [](http://example.com/codeflow?start=17&length=3)

ditto, return the type here.

mzasov replied:
fixed

tarekgh commented on src/Microsoft.Data.Analysis/DataFrame.Join.cs:
```diff
@@ -178,89 +182,121 @@ private static HashSet<long> Merge(DataFrame retainedDataFrame, DataFrame supple
+                // Convert indices from in key from local (shrinked row) to indices in original dataframe
                 if (retainedIndicesReverseMapping != null)
-                    newOccurrences = newOccurrences.ToDictionary(kvp => retainedIndicesReverseMapping[kvp.Key], kvp => kvp.Value);
+                    newOccurrences = newOccurrences.ToDictionary(x => retainedIndicesReverseMapping[x.Key], x => x.Value);
```
>x [](http://example.com/codeflow?start=65&length=1)

kvp is better name here than x.

mzasov replied:
fixed

tarekgh commented:
@mzasov thanks for submitting this PR. Could you please paste some benchmark numbers for before and after the change to see the gain we get from this change? [Benchmark Dotnet](https://benchmarkdotnet.org/articles/overview.html) can be used to do that.

mzasov commented on src/Microsoft.Data.Analysis/DataFrame.Join.cs:
```diff
@@ -368,22 +403,62 @@ public DataFrame Merge(DataFrame other, string[] leftJoinColumns, string[] right
-            //insert columns from left dataframe (this)
-            for (int i = 0; i < this.Columns.Count; i++)
+            PrimitiveDataFrameColumn<long> mapIndicesLeft = isLeftDataFrameRetained ? retainedRowIndices : supplementaryRowIndices;
+            PrimitiveDataFrameColumn<long> mapIndicesRight = isLeftDataFrameRetained ? supplementaryRowIndices : retainedRowIndices;
```
no actually need to do that

tarekgh replied:
You are going to do the conditions once if you use it,

mzasov replied:
yes, sure, I extracted the code that not changed in the cycle and I think that keeping code simple if we can - it's a good way (without introducing of value tuple).

mzasov commented:
@tarekgh I've added benchmark output here:

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19042.1586 (20H2/October2020Update)
Intel Core i7-9700K CPU 3.60GHz (Coffee Lake), 1 CPU, 8 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4420.0), X64 RyuJIT
  Job-OBEJVW : .NET Framework 4.8 (4.8.4420.0), X64 RyuJIT

IterationCount=1  LaunchCount=1  WarmupCount=1  

|                        Method |      Mean | Error |
|------------------------------ |----------:|------:|
|         TestGetOccurencesSlow | 547.371 s |    NA |
| TestGetOccurencesIntermediate |  55.985 s |    NA |
|         TestGetOccurencesFast |   3.422 s |    NA |

where TestGetOccurencesSlow - is the current version
TestGetOccurencesIntermediate - is the intermediate version with hash sets intersection
TestGetOccurencesFast - is the fixed version with intersection calculation for sorted lists

mzasov commented:
There is the benchmark result

[Microsoft.Data.Analysis.Tests.DataFrameTests.MergePerformanceTesting-20220408-154904.log](https://github.com/dotnet/machinelearning/files/8453975/Microsoft.Data.Analysis.Tests.DataFrameTests.MergePerformanceTesting-20220408-154904.log)

tarekgh reviewed (approved):
Thanks @mzasov for doing this change. LGTM

tarekgh commented:
One thing we need to validate before merging this PR is we guarantee the lists we are merging in the method `GetSortedListsIntersection` is always sorted.

mzasov commented:
@tarekgh I added some tests to cover new functionality and extract GetSortedListsIntersection method to extensions (to make tests possible). Also added comment to DataFrameColumn.GroupColumnValues - there is should be sorted result (and all implementations provide sorted collection of longs in the result).

michaelgsharp commented:
I am fine with making that a requirement @tarekgh. I would also bet it won't break anyone, and since the DataFrame is still in preview now is the time to make that change if we ever wanted to make it.

tarekgh commented:
Sounds good then. Now, we need to ensure the doc change gets published publicly.

tarekgh commented:
@mzasov thanks again for your effort. I am going to merge it. Do you mind opening a doc PR to add the extra comment require the collections be sorted? https://github.com/dotnet/ml-api-docs/blob/main/dotnet/xml/Microsoft.Data.Analysis/DataFrameColumn.xml
